src/crypto/crypto-ops.c

```diff
@@ -3831,14 +3831,39 @@ int sc_isnonzero(const unsigned char *s) {
 }
 
 int ge_p3_is_point_at_infinity(const ge_p3 *p) {
-  // X = 0 and Y == Z
-  int n;
-  for (n = 0; n < 10; ++n)
+  // https://eprint.iacr.org/2008/522
+  // X == T == 0 and Y/Z == 1
+  // note: convert all pieces to canonical bytes in case rounding is required (i.e. an element is > q)
+  char result_X_bytes[32];
+  char result_Y_bytes[32];
+  char result_Z_bytes[32];
+  char result_T_bytes[32];
+  fe_tobytes((unsigned char*)&result_X_bytes, p->X);
+  fe_tobytes((unsigned char*)&result_Y_bytes, p->Y);
+  fe_tobytes((unsigned char*)&result_Z_bytes, p->Z);
+  fe_tobytes((unsigned char*)&result_T_bytes, p->T);
+
+  // X != 0 || T != 0
+  for (int i = 0; i < 32; ++i)
   {
-    if (p->X[n] | p->T[n])
+    if (result_X_bytes[i] | result_T_bytes[i])
```

Comment:
Does `T` need to be evaluated at all? I think it can be skipped entirely in this function. Dropping the `fe_tobytes` for `T` should be a performance win.

The point at infinity should be `(x,y) = (0, 1)`, and `T` is defined as `(X*Y)/Z` (or `t=x*y`). So if `X` is `0`, then `T` must also be zero ... ?

Reply:
What you're saying sounds right, but I don't have the expertise to confidently agree. Hopefully we can get more opinions.

Reply:
One comment/question I got was whether skipping this test would be safe if given an input not produced by a 'correct' method.

Reply:
My understanding of the coordinates used in the reference [https://eprint.iacr.org/2008/522](url) is that X=0 is a sufficient condition for T=0 since Z != 0 by definition and therefore Y is bounded from above.  If the ``point at infinity'' is defined as (X:Y:Z:T)=(0:1:0:1) within the equivalence class of $\lambda$-rescalings, then the definition T=XY/Z, Z\neq 0, implies that checking X=0 and Y/Z=1 is enough. In other words, X=0 and Y/Z=1 is a sufficient condition for the point to a conformal representation of the identity in these coordinates.

Reply:
> The point at infinity should be (x,y) = (0, 1), and T is defined as (X*Y)/Z (or t=x*y). So if X is 0, then T must also be zero ... ? @vtnerd 

`T = XY/Z` is only true for valid point representations. If an invalid point representation is passed in (where `Y/Z == 1`, `X == 0`, `T != 0`), then this function should _not_ label it as the point-at-infinity. Here is some algebra (using equation 5 from [the paper](https://eprint.iacr.org/2008/522.pdf)) showing that adding an invalid point `P2` to a normal point `P1` will not produce `P1`: `P1 + P2 != P1`.

```
P1: {X1, Y1, T1, Z1}
P2 (invalid identity): {X2 = 0, Y2 != 0, T2 != 0, Z2 = Y2}

1. apply point-addition: P1 + P2 = P3
X3 = X1Y2*(Z1Y2 - dT1T2) = X1*(Z1Y2^2 - dT1T2Y2)
Y3 = Y1Y2*(Z1Y2 + dT1T2) = Y1*(Z1Y2^2 + dT1T2Y2)
T3 = Y1Y2*X1Y2           = X1Y1Y2^2
Z3 = (Z1Y2 - dT1T2)*(Z1Y2 + dT1T2) = (Z1Y2)^2 - (dT1T2)^2

2. multiply by 1/Y2
X3 = X1Y2*(Z1Y2 - dT1T2) = X1*(Z1Y2 - dT1T2)
Y3 = Y1Y2*(Z1Y2 + dT1T2) = Y1*(Z1Y2 + dT1T2)
T3 = Y1Y2*X1Y2           = X1Y1Y2
Z3 = (Z1Y2 - dT1T2)*(Z1Y2 + dT1T2) = (1/Y2) [ (Z1Y2)^2 - (dT1T2)^2 ]

3. check T-property
X3Y3 = X1Y1 [ (Z1Y2)^2 - (dT1T2)^2 ]
X3Y3/Z3 = X1Y1Y2 = T3

4. multiply by 1/Z1Y2
X3 = X1Y2*(Z1Y2 - dT1T2) = X1 - X1*dT1T2/Z1Y2
Y3 = Y1Y2*(Z1Y2 + dT1T2) = Y1 + Y1*dT1T2/Z1Y2
T3 = Y1Y2*X1Y2           = X1Y1/Z1
Z3 = (Z1Y2 - dT1T2)*(Z1Y2 + dT1T2) = Z1 - (dT1T2)^2/Z1Y2^2

If T2 == 0, then all the junk disappears and P3 == P1 as expected. Can you make the junk go away when T2 != 0?
```